internal/sql/pg/query.go

```diff
@@ -234,3 +235,92 @@ func queryTx(ctx context.Context, oidToDataType map[uint32]*datatype, dbTx txBeg
 
 	return resSet, err
 }
+
+func queryRowFunc(ctx context.Context, conn *pgx.Conn, sql string,
+	scans []any, fn func() error, args ...any) error {
+	rows, _ := conn.Query(ctx, sql, args...)
+	_, err := pgx.ForEachRow(rows, scans, fn)
+	return err
+}
+
+// QueryRowFunc will attempt to execute an SQL statement, handling the rows and
+// returned values as described by the sql.QueryScanner interface. If the
+// provided Executor is also a sql.QueryScanner, that method will be used,
+// otherwise, it will attempt to use the underlying DB connection. The latter is
+// supported for all concrete transaction types in this package as well as
+// instances of the pgx.Tx interface.
+func QueryRowFunc(ctx context.Context, tx sql.Executor, stmt string,
+	scans []any, fn func() error, args ...any) error {
+	switch ti := tx.(type) {
+	case sql.QueryScanner:
+		return ti.QueryScanFn(ctx, stmt, scans, fn, args...)
+	case conner:
+		conn := ti.Conn()
+		return queryRowFunc(ctx, conn, stmt, scans, fn, args...)
+	}
+	return errors.New("cannot query with scan values")
+}
+
+// QueryRowFuncAny is similar to QueryRowFunc, except that no scan values slice
+// is provided. The provided function is called for each row of the result. The
+// caller does not determine the types of the Go variables in the values slice.
+// In this way it behaves similar to Execute, but providing "for each row"
+// functionality so that every row does not need to be loaded into memory. See
+// also QueryRowFunc, which allows the caller to provide a scan values slice.
+func QueryRowFuncAny(ctx context.Context, tx sql.Executor, stmt string,
+	fn func([]any) error, args ...any) error {
+	conner, ok := tx.(conner)
+	if !ok {
+		return errors.New("no conn access")
+	}
+	conn := conner.Conn()
+	return queryRowFuncAny(ctx, conn, stmt, fn, args...)
+}
+
+func queryRowFuncAny(ctx context.Context, conn *pgx.Conn, stmt string,
+	fn func(vals []any) error, args ...any) error {
+	oidTypes := oidTypesMap(conn.TypeMap())
+
+	rows, _ := conn.Query(ctx, stmt, args...)
+	fields := rows.FieldDescriptions()
+	var oids []uint32
+	for _, f := range fields {
+		// NOTE: for a domain (constrained alias) this will be the OID of the underlying type
+		oids = append(oids, f.DataTypeOID)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		pgxVals, err := rows.Values()
+		if err != nil {
+			return err
+		}
+
+		// Decode the values into Kwil or native types.
+		decVals := make([]any, len(pgxVals))
+		for i, pgVal := range pgxVals {
+			decVal, err := decodeFromPGVal(pgVal, oids[i], oidTypes)
```

Comment:
Since I  moved `queryRowFuncAny` to a new file, I'll call out this key change, which is converting the types to Kwil types according to our oid type map.